atomate/vasp/workflows/base/magnetism.py

```diff
@@ -0,0 +1,834 @@
+        if "ferrimagnetic_by_motif" in self.strategies and len(wyckoff_symbols) > 1:
+
+            # these orderings are AFM on one local environment, and FM on the rest
+            for symbol in wyckoff_symbols:
+
+                constraints = [
+                    MagOrderParameterConstraint(
+                        0.5, site_constraint_name="wyckoff", site_constraints=symbol
+                    ),
+                    MagOrderParameterConstraint(
+                        1.0,
+                        site_constraint_name="wyckoff",
+                        site_constraints=list(wyckoff_symbols - {symbol}),
+                    ),
+                ]
+
+                all_constraints["ferri_by_motif_{}".format(symbol)] = constraints
+
+        # and also try ferrimagnetic when there are multiple magnetic species
+        if "ferrimagnetic_by_species" in self.strategies:
+
+            sp_list = [str(site.specie) for site in structure]
+            num_sp = {sp: sp_list.count(str(sp)) for sp in types_mag_species}
+            total_mag_sites = sum(num_sp.values())
+
+            for sp in types_mag_species:
+
+                # attempt via a global order parameter
+                all_constraints["ferri_by_{}".format(sp)] = num_sp[sp] / total_mag_sites
+
+                # attempt via afm on sp, fm on remaining species
+
+                constraints = [
+                    MagOrderParameterConstraint(0.5, species_constraints=str(sp)),
+                    MagOrderParameterConstraint(
+                        1.0,
+                        species_constraints=list(
+                            map(str, set(types_mag_species) - {sp})
+                        ),
+                    ),
+                ]
+
+                all_constraints["ferri_by_{}_afm".format(sp)] = constraints
+
+        # ...and finally, we can try orderings that are AFM on one local
+        # environment, and non-magnetic on the rest -- this is less common
+        # but unless explicitly attempted, these states are unlikely to be found
+        if "antiferromagnetic_by_motif" in self.strategies:
+
+            for symbol in wyckoff_symbols:
+
+                constraints = [
+                    MagOrderParameterConstraint(
+                        0.5, site_constraint_name="wyckoff", site_constraints=symbol
+                    )
+                ]
+
+                all_constraints["afm_by_motif_{}".format(symbol)] = constraints
+
+        # and now construct all our transformations for each strategy
+        transformations = {}
+        for name, constraints in all_constraints.items():
+
+            trans = MagOrderingTransformation(
+                mag_species_spin,
+                order_parameter=constraints,
+                **self.transformation_kwargs
+            )
+
+            transformations[name] = trans
+
+        return transformations
+
+    def _generate_ordered_structures(self, sanitized_input_structure, transformations):
```

Comment:
See my comment regarding the previous function and pymatgen vs atomate

Ideally, atomate routines should be short and just glue together existing functionality in pymatgen into a set of Fireworks and Workflow objects. i.e. call a bunch of pymatgen functions and wrap that into a Firework/Workflow object.

When a lot of algorithm work is put in atomate, two problems occur:
* As atomate is less used than pymatgen, fewer people and developers review, vet, and improve the algorithms
* Generally, tying together workflows is confusing enough code for people to follow as workflows are somehow inherently confusing. When you mix in algorithm work together with the workflow stuff (even if the algorithms are in a separate private function), wrapping your head around workflows becomes even harder. So to keep atomate workflows easy to maintain, it's best if they are as short as possible and use pymatgen for most of the heavy lifting

Reply:
Updated comment above about this. I did add a lot of underlying improvements to pymatgen too, it's just that some extra glue was needed to generate a plausible set of inputs to these updated transformations, and it *felt* like this was more a workflow thing than a pymatgen thing, but that is a judgement call.